npe2/manifest/io.py

```diff
@@ -15,3 +16,193 @@ class ReaderContribution(BaseModel):
     accepts_directories: bool = Field(
         False, description="Whether this reader accepts directories"
     )
+
+    class Config:
+        extra = Extra.forbid
+
+
+class LayerType(str, Enum):
+    image = "image"
+    labels = "labels"
+    points = "points"
+    shapes = "shapes"
+    surface = "surface"
+    tracks = "tracks"
+    vectors = "vectors"
+
+
+class LayerTypeConstraint(BaseModel):
+    """Layer type constraints.
+
+    A writer plugin can declare that it will write 0 or more layers of a
+    specific type.
+
+    For example:
+
+    ```
+        image      Write exactly 1 image layer.
+        image?     Write 0 or 1 image layes.
+        image+     Write 1 or more image layers.
+        image*     Write 0 or more image layers.
+        image{k}   Write exactly k image layres.
+        image{m,n} Write between m and n layers (inclusive range). Must have m<=n.
+    ```
+
+    When a type is not present in the list of constraints, that
+    corresponds to a writer that is not compatible with that type.
+
+    For example, a writer declaring:
+
+    ```
+        layer_types=["image+", "points*"]
+    ```
+
+    would not be selected when trying to write an `image` and a `vector`
+    layer because the above only works for cases with 0 `vector` layers.
+
+    Note that just because a writer declares compatibility with a layer
+    type does not mean it actually writes that type.  In the example
+    above, the writer might accept a set of layers containing `image`s and
+    `point`s, but the write command might just ignore the `point` layers
+    """
+
+    layer_type: LayerType
+    bounds: Tuple[int, Optional[int]] = Field(
```

Comment:
I seems that `bounds[1]` will _always_ be an int right?  (I don't see anywhere where it ends up being `None`).  Furthermore, I get a type error if I try `IntervalTree.addi(1, None, '...')` ... so it seems like this needs to be `Tuple[int, int]`,  yeah?